stablehlo/dialect/TypeInference.cpp

```diff
@@ -2464,28 +2464,35 @@ LogicalResult inferPadOp(std::optional<Location> location, Value operand,
   auto inputType = operand.getType().cast<RankedTensorType>();
   auto padType = paddingValue.getType().cast<RankedTensorType>();
 
+  // pad_i2
   if (padType.getRank() != 0)
     return emitOptionalError(location,
                              "padding value type should be a rank-0 "
                              "tensor, is rank ",
                              padType.getRank());
 
+  // pad_c2, pad_i3, pad_i4, pad_i5
+  if (edgePaddingLow.getType() != edgePaddingHigh.getType() ||
+      edgePaddingLow.getType() != interiorPadding.getType())
+    return emitOptionalError(
+        location, "edge_padding_low, edge_padding_high, ",
+        "and interior_padding must have the same type but got: ",
+        edgePaddingLow.getType(), ", ", edgePaddingHigh.getType(), ", and ",
+        interiorPadding.getType());
```

Comment:
Since we're refactoring this, looks like this check can be expressed via `AllTypesMatch` in the ODS? It'll reduce the amount of C++ code and will allow us to avoid writing some tests.

Reply:
I originally added the `AllTypesMatch` Trait, but @zhouxin913 suggested not to use this since dynamism work will not be compatible with this Trait.

Reply:
In this specific case, it'll be fine because `edge_padding_low`, `edge_padding_high` and `interior_padding` are a special case - they are not going to have dynamic types.